apisix/core/config_etcd.lua

```diff
@@ -157,45 +369,71 @@ local function flush_watching_streams(self)
 end
 
 
-local function http_waitdir(etcd_cli, key, modified_index, timeout)
-    local opts = {}
-    opts.start_revision = modified_index
-    opts.timeout = timeout
-    opts.need_cancel = true
-    local res_func, func_err, http_cli = etcd_cli:watchdir(key, opts)
-    if not res_func then
-        return nil, func_err
+local function http_waitdir(self, etcd_cli, key, modified_index, timeout)
+    if not watch_ctx.idx[key] then
+        watch_ctx.idx[key] = 1
     end
 
-    -- in etcd v3, the 1st res of watch is watch info, useless to us.
-    -- try twice to skip create info
-    local res, err = res_func()
-    if not res or not res.result or not res.result.events then
-        res, err = res_func()
-    end
+    ::iterate_events::
+    for i = watch_ctx.idx[key], #watch_ctx.res do
+        watch_ctx.idx[key] = i + 1
 
-    if http_cli then
-        local res_cancel, err_cancel = etcd_cli:watchcancel(http_cli)
-        if res_cancel == 1 then
-            log.info("cancel watch connection success")
-        else
-            log.error("cancel watch failed: ", err_cancel)
+        local item = watch_ctx.res[i]
+        if item == false then
+            goto iterate_events
+        end
+
+        local res, err = item.res, item.err
+        if err then
+            return res, err
+        end
+
+        local found = false
+        -- ignore res with revision smaller then self.prev_index
+        if tonumber(res.result.header.revision) > self.prev_index then
+            for _, evt in ipairs(res.result.events) do
+                if evt.kv.key:find(key) == 1 then
+                    found = true
+                    break
+                end
+            end
+        end
+
+        if found then
+            local res2 = tablex.deepcopy(res)
+            table.clear(res2.result.events)
+            for _, evt in ipairs(res.result.events) do
+                if evt.kv.key:find(key) == 1 then
+                    insert_tab(res2.result.events, evt)
+                end
+            end
+            if log_level >= NGX_INFO then
+                log.info("http_waitdir: ", inspect(res2))
+            end
+            return res2
         end
```

Comment:
```suggestion
        local events = {}
        if tonumber(res.result.header.revision) > self.prev_index then
            for _, evt in ipairs(res.result.events) do
                if evt.kv.key:find(key) == 1 then
                    insert_tab(events, evt)
                end
            end
        end

        if not core.table.isempty(events) then
            local res2 = tablex.deepcopy(res)
            res2.result.events = events
            return res2
        end
```

Reply:
No need to create a new table (and replace a newly created sub-table) here.

Reply:
```suggestion
        local found = false
        local res2
        if tonumber(res.result.header.revision) > self.prev_index then
            for _, evt in ipairs(res.result.events) do
                if evt.kv.key:find(key) == 1 then
                    if not found then 
                        res2 = tablex.deepcopy(res)
                        table.clear(res2.result.events)
                        found = true
                    end

                    insert_tab(res2.result.events, evt)
                end
            end
        end
```

Reply:
we don't need to iterate from the beginning of the `res.result.events` again

Reply:
fix.